### ISSUE TITLE
3DS - Player can't load game files from romFs (regression)

### ISSUE THREAD

**Issue by msikma** [EasyRPG/Player#2256]
#### Name of the game:

Any game

#### Player platform:

3DS

#### Attach files (as a .zip archive or link them):

* [Zip file of my build setup](http://wedemandhtml.com/tmp/cia_builder_ffxat.zip) (77MB) containing the game, the script, several `easyrpg-player.elf` files. If you have the required command line tools installed you can just run the build.sh script.

* Someone else [noticed this bug](https://community.easyrpg.org/t/problems-with-the-creation-of-a-custom-cia/804) a while back and posted about it on the forum.

* EasyRPG log below.

#### Other details

* 3dstool 1.2.3
* CTR MAKEROM v0.15
* Building on Mac OS X 10.14.6 Mojave
* Running on an old 3DS (tested on Citra as well)

#### Describe the issue in detail and how to reproduce it:

Hi everyone. It seems that there's a problem locating game files on 3DS, making it impossible to build self-contained CIA files that contain a single game.

This relates to building 3DS CIA files using the [CIA builder](https://github.com/EasyRPG/Player/tree/master/builds/3ds/cia_builder).

It seems this broke at some point: a much older `easyrpg-player.elf` file (build from 2017-02-11, unsure what the build number is) I have works fine and generates working CIA files (this version happens to look a little glitchy but it does work fine):

![easyrpg-20170211](https://user-images.githubusercontent.com/1507745/85866587-e8c7bb00-b7c7-11ea-9531-8152ae0cb955.png)

Today, with the latest version, building the same game in the exact same way leads to this error:

![easyrpg-770](https://user-images.githubusercontent.com/1507745/85866814-478d3480-b7c8-11ea-9a98-f1a0d7e0124e.png)

Note: I've tested both versions on a real 3DS and on Citra, and they both do the same thing (ignore the missing audio firmware complaint, I just haven't set that up in Citra yet, but I have it on 3DS and it doesn't make a difference whether it's there or not).

I happened to have two older versions: build 576, which also fails in the exact same way, and build 276 which unfortunately [crashes](https://user-images.githubusercontent.com/1507745/85866890-668bc680-b7c8-11ea-9b76-7a63cca2fe96.png) (both a real 3DS and Citra) before I can tell if it works or not.

I'm not able to find old builds of the 3DS `easyrpg-player.elf` file on the [Jenkins](https://ci.easyrpg.org/job/player-3ds/) page so unfortunately I can't track down where this bug appeared. Is there some other place where old builds are available? On Jenkins they only go back one week.

When I run my built CIA file on 3DS using the latest build, 770, the following log is generated (exiting almost immediately):

```
[2020-06-26 11:29:40] Debug: EasyRPG Player 0.6.2 (git~176@620278ca, 2020-06-25) started
[2020-06-26 11:29:40] Debug: ===========================================================
[2020-06-26 11:29:40] Debug: Clock: CtrClock steady=true period=custom (1 / 268123480)
[2020-06-26 11:29:40] Debug: Seeded the RNG with 1593170980.
[2020-06-26 11:29:40] Debug: CLI: 
[2020-06-26 11:29:40] Debug: Detected a project on romFs filesystem...
[2020-06-26 11:29:40] Error: romfs: is not a valid path
```

When I run the older, working CIA file, I get the following log (starting a new game and playing for a bit, then exiting):

```
[2020-06-26 11:33:21] Debug: EasyRPG Player 0.5.0 (git~122@9ab6ac2, 2017-02-11) started
[2020-06-26 11:33:22] Debug: ==========================================================
[2020-06-26 11:33:22] Debug: csnd:SND has been selected as audio service.
[2020-06-26 11:33:22] Debug: CLI: 
[2020-06-26 11:33:22] Debug: Detected a project on romFs filesystem...
[2020-06-26 11:33:22] Debug: Detected encoding: UTF-8
[2020-06-26 11:33:22] Debug: Using romfs: as Game directory
[2020-06-26 11:33:22] Debug: Using sdmc:/easyrpg-player/00040000FFA03800 as Save directory
[2020-06-26 11:33:23] Debug: Loading game FF-XAT
[2020-06-26 11:33:23] Debug: Using RPG2k Interpreter
[2020-06-26 11:33:23] Debug: RPG2k < v1.50 / RPG2k3 < v1.05 detected
[2020-06-26 11:33:23] Debug: Engine configured as: 2k=1 2k3=0 2k3Legacy=0 MajorUpdated=0 2k3E=0
[2020-06-26 11:33:24] Warning: Unsupported music format (romfs:/Music/FF5cryst.mid)
[2020-06-26 11:33:32] Debug: Starting new game
[2020-06-26 11:33:32] Debug: Loading Map Map0042.lmu
[2020-06-26 11:33:32] Debug: Tree: Title
[2020-06-26 11:33:32] Warning: Unsupported music format (romfs:/Music/FF5fierc.mid)
[2020-06-26 11:33:59] Debug: Loading Map Map0043.lmu
[2020-06-26 11:33:59] Debug: Tree: Space ship Raknarok < World Map
[2020-06-26 11:33:59] Warning: Unsupported music format (romfs:/Music/castle2.mid)
[2020-06-26 11:34:05] Warning: Unsupported music format (romfs:/Music/FF5cryst.mid)
```

Interesting is that the newer build logs to `/easyrpg_log.txt` and the older build logs to `/easyrpg-player/00040000FFA03800/easyrpg_log.txt`, incorporating the game ID.

Note: I don't have EasyRPG installed on my 3DS anywhere, nor do I have an RTP anywhere. I deleted everything for testing. The game I linked above is self contained.

I hope someone can figure out what's going on, I think the 3DS is one of the most awesome platforms for running EasyRPG. 🙂

If anyone knows where I can find older builds of `easyrpg-player.elf`, I will go and track down where this broke more precisely.

**Comment by Ghabry** [comment]
Thanks for the report!
We are not sure if we broke this or ctrulib.
Thanks for finding a known good version. Helps us to find the culprit (by compiling this old version with new libraries)

**Comment by Ghabry** [comment]
@msikma 
Is a 3dsx or an existing cia suitable for building one with a romfs?
We don't keep the elfs around only the others: https://easyrpg.org/downloads/player/

**Comment by msikma** [comment]
I'm not entirely sure. I think the build instructions explicitly require an elf file but maybe there's a way to convert the 3dsx file to it. I'll have to ask around to see how exactly elf/3dsx files compare. I'm pretty sure cia files won't work since as I understand it they're like 3dsx files but with some extra metadata to make them suitable for installing on a 3DS home screen.

If not I'll try to see if I can make some builds myself to test with.

**Comment by WaffleGM** [comment]
Any updates on this issue or any other way to obtain the files OP used? The download doesn't seem to be working on my end anymore. I've been trying to port some games myself for a friend, but am coming across this same issue.

**Comment by carstene1ns** [comment]
There was an issue with maybe `stat` (do not remember correctly anymore) to detect romfs folders, so our existence check failed.
Needs testing, if this is still the case.

@WaffleGM I would not recommend to use 5 year old binaries today. There have been much improvements in the player overall since then.

**Comment by carstene1ns** [comment]
Apparently there is more wrong with it, as simply fixing this problem makes player hang shortly after.
Debugging with citra is a pain, need to test this with my real 3DS.

```patch
diff --git a/src/main_data.cpp b/src/main_data.cpp
index 7ae2a5c2..4dd0ee5e 100644
--- a/src/main_data.cpp
+++ b/src/main_data.cpp
@@ -47,7 +47,8 @@
 #  include <SDL_system.h>
 #elif defined(__3DS__)
 #  include <3ds.h>
-#  include <cstdio>
+#  include <unistd.h>
+   extern int __system_argc; // inside ctrulib
 #elif defined(__vita__)
 #  include <cstdio>
 #  include <psp2/io/stat.h>
@@ -114,40 +115,38 @@ void Main_Data::Init() {
 				FileFinder::SetSaveFilesystem(savefs);
 			}
 #elif defined(__3DS__)
-			// Check if romFs has some files inside or not
-			FILE* testfile = fopen("romfs:/RPG_RT.lmt","r");
-			if (testfile != NULL) {
-				Output::Debug("Detected a project on romFs filesystem...");
-				fclose(testfile);
-				project_path = "romfs:";
+			bool is_cia = __system_argc == 0;
+			char tmp_path[64] = "sdmc:/3ds/easyrpg-player";
 
-				if (!Player::is_3dsx) {
-					// Create savepath for CIA - unique for any ID
+			// Check if romfs has some files inside or not
+			if(::access("romfs:/RPG_RT.lmt", F_OK) == 0) {
+				Output::Debug("Loading game from romfs...");
+				chdir("romfs:/");
 
-					// Generating save_path
+				if (is_cia) {
+					// CIA savepath is unique for any ID
 					u64 titleID;
 					APT_GetProgramID(&titleID);
-					char mainDir[64];
-					sprintf(mainDir,"sdmc:/easyrpg-player/%016llX",titleID);
+					sprintf(tmp_path, "sdmc:/easyrpg-player/%016llX", titleID);
 
 					// Creating dirs if they don't exist
 					FS_Archive archive;
 					FSUSER_OpenArchive(&archive, ARCHIVE_SDMC, {PATH_EMPTY, 1, (u8*)""});
 					FS_Path filePath=fsMakePath(PATH_ASCII, "/easyrpg-player");
 					FSUSER_CreateDirectory(archive,filePath, FS_ATTRIBUTE_DIRECTORY);
-					FS_Path filePath2=fsMakePath(PATH_ASCII, &mainDir[5]);
+					FS_Path filePath2=fsMakePath(PATH_ASCII, &tmp_path[5]);
 					FSUSER_CreateDirectory(archive,filePath2, FS_ATTRIBUTE_DIRECTORY);
 					FSUSER_CloseArchive(archive);
+				}
 
-					auto savefs = FileFinder::Root().Create(mainDir);
-					if (!savefs) {
-						Output::Error("3DS: Invalid save path {}", mainDir);
-					}
-					FileFinder::SetSaveFilesystem(savefs);
+				auto savefs = FileFinder::Root().Create(tmp_path);
+				if (!savefs) {
+					Output::Error("3DS: Invalid save path {}", tmp_path);
 				}
-			} else if (!Player::is_3dsx) {
+				FileFinder::SetSaveFilesystem(savefs);
+			} else if(is_cia) {
 				// No RomFS -> load games from hardcoded path
-				project_path = "sdmc:/3ds/easyrpg-player";
+				project_path = tmp_path;
 			}
 #elif defined(__APPLE__) && TARGET_OS_OSX
 			// Apple Finder does not set the working directory
diff --git a/src/player.cpp b/src/player.cpp
index 5c587af8..ad863ee7 100644
--- a/src/player.cpp
+++ b/src/player.cpp
@@ -134,9 +134,6 @@ namespace Player {
 #ifdef EMSCRIPTEN
 	std::string emscripten_game_name;
 #endif
-#ifdef __3DS__
-	bool is_3dsx;
-#endif
 }
 
 namespace {
@@ -442,10 +439,6 @@ void Player::Exit() {
 }
 
 Game_Config Player::ParseCommandLine(int argc, char *argv[]) {
-#ifdef __3DS__
-	is_3dsx = argc > 0;
-#endif
-
 #if defined(_WIN32) && !defined(__WINRT__)
 	int argc_w;
 	LPWSTR *argv_w = CommandLineToArgvW(GetCommandLineW(), &argc_w);
diff --git a/src/player.h b/src/player.h
index c108b6c1..a539f90a 100644
--- a/src/player.h
+++ b/src/player.h
@@ -373,11 +373,6 @@ namespace Player {
 	/** Name of game emscripten uses */
 	extern std::string emscripten_game_name;
 #endif
-
-#ifdef __3DS__
-	/** Is executed from a .3dsx (otherwise .cia) */
-	extern bool is_3dsx;
-#endif
 }
 
 inline bool Player::IsRPG2k() {
```

**Comment by Ghabry** [comment]
Hmm this could be a bug in the filesystem code. Likely some small difference in how romfs:/ works compared to sdmc:/

**Comment by msikma** [comment]
I'm really glad that there's still interest in looking into this! The 3DS is a really nice platform for playing Rm2k games I think. It would be cool if it could someday be fixed, even though the 3DS is now a pretty underused platform.

It has been a while since I've tried compiling EasyRPG but I'll do so soon to see if I can help test any updates. Would be nice to be able to update [my old EasyRPG 3DS build script](https://github.com/msikma/Easy3DS).

**Comment by Ghabry** [comment]
This already crashes on Linux when doing ``--project-path /`` so is likely a good place to start the investigation :)

**Comment by carstene1ns** [comment]
> crashes on Linux when doing `--project-path /`

This does not happen here.

**Comment by carstene1ns** [comment]
Putting a game in a subfolder on romfs gives you the game browser and from there it loads fine.

One thing missing in my test was that you would need an actual game in `/`, so scratch my previous comment.

**Comment by carstene1ns** [comment]
While debugging, `FileFinder::GetPathAndFilename()` was interesting.

If we look at the test, we see that a file in cwd and in filesystem root is treated very similar:
Sure, this is an edge case, however such special handling of `/` might be cluttered in other places.
https://github.com/EasyRPG/Player/blob/a079b895c7988a51ddd3d657d1f72be47df7e5f3/tests/filefinder.cpp#L63-L69

**Comment by Ghabry** [comment]
@msikma  This should work in the latest version. At least we hope that it works

**Comment by Ghabry** [comment]
Btw the screenshot by c1 shows nicely what is likely happening on the Vita (so is probably the same bug):

It goes from ``ListDirectory: romfs:`` to ``ListDirectory: [empty string]``. Then this is interpreted as ``.`` and _boom_.

That bug is now kinda gone because a ``/`` is always appended but assuming the drive is completely invalid the code should stop going "upwards" when a drive letter is encountered.

**Comment by carstene1ns** [comment]
@Ghabry you forgot the original bug is still present, one would at least need to change "romfs:" to "romfs:/" in main_data.cpp

**Comment by Ghabry** [comment]
Oh you are right. Didn't care about the main_data.cpp... So this is technically not even fixed :-1:

**Comment by carstene1ns** [comment]
Fixed by #2792.
